source/puppet/4.4/reference/static_catalogs.md

```diff
@@ -0,0 +1,132 @@
+
+> **Note:** The Puppet Server process must be able to execute these scripts. Puppet Server also validates their output and checks their exit codes. Environment names can contain only alphanumeric characters and underscores (`_`). The `code_id` can  contain only alphanumeric characters and dashes (`-`), underscores (`_`), semicolons (`;`), and colons (`:`). If either command returns a non-zero exit code, Puppet Server logs an error and returns the error message and a 500 response code to the API request.
+
+Puppet Server validates the standard output of each of these scripts, and if the output's acceptable, it adds the results to the catalog as their respective parameters' values. This lets you use any versioning or synchronization tools you want, as long as you can write scripts that produce a valid string for the `code_id` and code content using the catalog's `code_id` and file's environment.
+
+The `code-id-command` and `code-content-command` scripts can be as simple or complex as necessary.
+
+> **Note:** The following examples demonstrate how Puppet Server passes arguments to the `code-id-command` and `code-content-command` scripts and how Puppet Server uses the results to return a specific version of a file resource. The scripts themselves are not recommended for production use.
+
+For instance, if files in an environment are simply managed by git, a `code-id-command` script can be nothing more than this, with the environment name passed to the script as the first command-line argument:
+
+``` bash
+#!/bin/bash
+set -e
+if [[ -z $1 ]]; then
+  echo Expected an environment >&2
+  exit 1
+fi
+cd /etc/puppetlabs/code/environments/"$1" && git rev-parse HEAD
+```
+
+As long as the script's exit code is zero, Puppet Server uses the script's standard output as the catalog's `code_id`.
+
+A `code-content-command` script can also be simple. Puppet Server passes the environment name as the first command-line argument, the `code_id` as the second argument, and the path to the file resource from its `content_uri` as the third argument:
+
+``` bash
+#!/bin/bash
+set -e
+if [[ $# < 3 ]]; then
+  echo Expected environment, code-id, file-path >&2
+  exit 1
+fi
+cd /etc/puppetlabs/code/environments/"$1" && git show $2:$3
+```
+
+The script's standard output is then provided as the file's `code_content` as long as the script returns a non-zero exit code.
+
+### Enabling or disabling static catalogs
+
+If you're using static catalogs, the agents don't need to request [file metadata][] or recurse into directories. And since static catalogs allow agents to use static catalogs more reliably, they're less likely to need to request catalogs as frequently.
+
+In other words, even if you aren't using static catalogs, disabling it doesn't substantially improve server or agent performance.
+
+However, if you still want to toggle static catalog generation, you can do so with the Boolean `static_catalogs` setting in two places:
+
```

Comment:
Sorry I think I mixed positive and negatives in my points. Let me update those.

Longer story, puppet has a built-in `modules` mount point (along with `plugins`, `pluginsync`, etc). Users can also create their own mount points to serve other files. So point 3 is to make sure the request was for the `modules` mount point specifically.

If the request gets past that check, then we try to resolve where the actual file is on the master (to compute checksum, etc). Normally the pattern is a request for `source => 'puppet:///modules/foo/bar'` will resolve to `/etc/puppetlabs/code/environments/<environment>/modules/foo/files/bar`, and in that case we would inline the resource.

However, due to symlinks (assuming opensource puppetserver with no file sync), it's possible that the file lives outside of the per-environment directory. So that's a case where we'd get past point 3, but fail on 4, and not inline metadata.

Also it's possible to have an `environment.conf` that [modifies the `modulepath`](https://github.com/puppetlabs/control-repo/blob/production/environment.conf#L1). If the resolved path is something like `/etc/puppetlabs/code/environments/<environment>/dist/foo/files/bar` then we would inline, even though `modules` is not in the file path.

It's also possible for `modulepath` in `environment.conf` to be something like `/usr/modules`, and in that case we would not inline.